src/clone.ts

```diff
@@ -1,68 +1,110 @@
-/* eslint-disable eqeqeq, guard-for-in, @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-return, @typescript-eslint/prefer-destructuring, @typescript-eslint/prefer-readonly-parameter-types, unicorn/no-null -- FIXME! */
+/* eslint-disable @typescript-eslint/prefer-readonly-parameter-types --
+ * The state needed to compute the clone is passed by reference via mutable
+ * arrays.
+ */
+
+import { purry } from "./purry";
+
+/**
+ * Creates a deep copy of the value. Supported types. Functions are assigned by
+ * reference rather than copied.
+ *
+ * @param data - The object to clone.
+ * @signature
+ *   R.clone(data)
+ * @example
+ *   R.clone({foo: 'bar'}) // {foo: 'bar'}
+ * @dataFirst
+ * @category Object
+ */
+export function clone<T>(data: T): T;
 
 /**
- * Creates a deep copy of the value. Supported types: `Array`, `Object`, `Number`, `String`, `Boolean`, `Date`, `RegExp`. Functions are assigned by reference rather than copied.
+ * Creates a deep copy of the value. Supported types. Functions are assigned by
+ * reference rather than copied.
  *
- * @param value - The object to clone.
- * @signature R.clone(value)
- * @example R.clone({foo: 'bar'}) // {foo: 'bar'}
+ * @signature
+ *   R.clone()(data)
+ * @example
+ *   R.pipe({foo: 'bar'}, R.clone()) // {foo: 'bar'}
+ * @dataLast
  * @category Object
  */
-export function clone<T>(value: T): T {
-  return value != null && typeof (value as any).clone === "function"
-    ? (value as any).clone()
-    : cloneImplementation(value, [], []);
+export function clone(): <T>(data: T) => T;
+
+export function clone(...args: ReadonlyArray<unknown>): unknown {
+  return purry(cloneImplementation, args);
 }
 
 // from https://github.com/ramda/ramda/blob/master/source/internal/_clone.js
-function cloneImplementation(
-  value: any,
-  refFrom: Array<any>,
-  refTo: Array<any>,
-): unknown {
-  function copy(copiedValue: any): unknown {
-    const len = refFrom.length;
-    let idx = 0;
-    while (idx < len) {
-      if (value === refFrom[idx]) {
-        return refTo[idx];
-      }
-      idx += 1;
-    }
-    refFrom[idx + 1] = value;
-    refTo[idx + 1] = copiedValue;
-    for (const key in value) {
-      copiedValue[key] = cloneImplementation(value[key], refFrom, refTo);
-    }
-    return copiedValue;
+function cloneImplementation<T>(
+  value: T,
+  refFrom: Array<unknown> = [],
+  refTo: Array<unknown> = [],
+): T {
+  if (typeof value === "function") {
+    // Functions aren't cloned, we return the same instance.
+    return value;
+  }
+
+  if (
+    typeof value !== "object" ||
+    value === null ||
+    value instanceof Date ||
+    value instanceof RegExp
+  ) {
+    // We can use the built-in deep cloning function.
+    return structuredClone(value);
```

Comment:
**thought:** for some reason i thought structuredClone had bad compat but it's fine

Reply:
Our tests run on Node 18 too, so if they pass i'm going to assume that the support is adequete for our use-case.